Nodejs/Product/Nodejs/SharedProject/HierarchyIdMap.cs

```diff
@@ -30,24 +34,21 @@ public uint Add(HierarchyNode node)
                 }
             }
 #endif
-
-            VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread();
-
-            lock (this.theLock)
+            if (!this.freedIds.TryPop(out var idx))
             {
-                if (this.freedIds.Count > 0)
-                {
-                    var i = this.freedIds.Pop();
-                    this.ids[i] = new WeakReference<HierarchyNode>(node);
-                    return (uint)i + 1;
-                }
-                else
-                {
-                    this.ids.Add(new WeakReference<HierarchyNode>(node));
-                    // ids are 1 based
-                    return (uint)this.ids.Count;
-                }
+                idx = this.NextIndex();
             }
+
+            var addSuccess = this.nodes.TryAdd(idx, new WeakReference<HierarchyNode>(node));
+            Debug.Assert(addSuccess, "Failed to add a new item");
+
+            return idx;
+        }
+
+        private uint NextIndex()
```

Comment:
>private uint NextIndex() [](start = 8, length = 24)

Since this is only correct if the free-list is empty, I would prefer to have it inlined.